### ISSUE TITLE
Terraform DevOps pipeline example

### ISSUE THREAD

**Issue by patheard** [dto-btn/todo#93]
Use `hashicorp/terraform:light` to plan and apply simple Terraform configuration.

[Examples of commands](https://blog.elreydetoda.site/docker-terraform/):

# With project ID and auth token
```bash
# first do an init to pull down plugins
docker run -it --rm -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light init /terraform/

# then do a plan out to a file
docker run -it --rm -e TF_VAR_project_id -e TF_VAR_auth_token -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light plan -out /terraform/main.plan /terraform/

# apply the plan
docker run -it --rm -e TF_VAR_project_id -e TF_VAR_auth_token -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light apply -state /terraform/main.tfstate /terraform/main.plan

# lastly destroy everything
docker run -it --rm -e TF_VAR_project_id -e TF_VAR_auth_token -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light destroy -auto-approve -state /terraform/main.tfstate /terraform/
```

# No TF_VARs
```bash
# first do an init to pull down plugins
docker run -it --rm -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light init /terraform/

# then do a plan out to a file
docker run -it --rm -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light plan -out /terraform/main.plan /terraform/

# apply the plan
docker run -it --rm -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light apply -state /terraform/main.tfstate /terraform/main.plan

# lastly destroy everything
docker run -it --rm -v "$(pwd)/.terrform":/.terraform/ -v "$(pwd)":/terraform hashicorp/terraform:light destroy -auto-approve -state /terraform/main.tfstate /terraform/
```

**Comment by patheard** [comment]
* Initial pipeline with terraform init/validate added.
* Next step is to add deploy to dev.
* Should also define stages in main .gitlab-ci.yml for consistency.

**Comment by patheard** [comment]
Pipeline created for AzureStack Deployments repo. Includes example of how to create jobs for specific folders in a monorepo as well.

**Comment by patheard** [comment]
`only:changes` does not work as expected when pushing a new branch:

https://gitlab.com/gitlab-org/gitlab/-/issues/27875

**tl;dr;**

On a new branch push, all files are considered changed.

**Comment by patheard** [comment]
Fixed by:

* limiting pipelines to only run for merge_requests and master.
* making the deploy jobs manual and optional.

Only side effect is that now there are two skipped pipelins for the DEV deploy job:

![image](https://user-images.githubusercontent.com/2110107/81715094-c9f2ac80-9445-11ea-8c63-e0506ed1892d.png)
